Move from LDLib's FluidStack implementation to Forge's FluidStack

krossgg opened the pull request:
## What
This PR moves the codebase off of LDLib's FluidStack implementation and onto Forge's native FluidStack implementation.
This includes switching from LDLib's IFluidTransfer to Forge's IFluidHandler. The behavior between the two is extremely similar, and this change will allow for better maintainability and better understanding of the depended-upon implementation, as Forge is very well documented.

## Implementation Details
I have added a couple new methods to `GTUtils` which I use sporadically, please see them [here](https://gist.github.com/krossgg/ff0a2496a195257dfd8e3e88782708dd) so you don't have to go flopping around trying to find what a certain utility method will do.
I also make very good use of Forge's `FluidUtil` class, which has many helper methods similar to LDLib's. See the docs for those [here](https://mcstreetguy.github.io/ForgeJavaDocs/1.20.1-47.2.0/net/minecraftforge/fluids/FluidUtil.html) or check the implementation in your own IDE at `net.minecraftforge.fluids.FluidUtil.java`

## Potential Compatibility Issues
Forge's FluidStacks are limited to _integer_ amounts, rather than the long-based FluidStacks from LDLib. This _could_ potentially cause issues in certain cases, specifically during interactions with AE2. 
Further testing is required for this, however. From what I can tell, it would only affect people who attempt to use more than 2.147 million buckets in an ME hatch or an ME Pattern Buffer.
If this is a cause for concern, please let me know.


### **If I have missed ANYTHING or if there are ANY questions please let me know.**
The first commit `4ef64c9` are fairly simple switches and don't necessarily need the most attention. Please look over the other four commits well. Thanks.

krossgg committed:
Required changes to integrate with LDLib

Adds classes in /api/misc/lib/ which are Forge FluidStack compatible stand-ins for LDLib's versions
Creates a Payload for Forge FluidStacks and registers them with LDLib
Adds a mixin for LDLib to allow for FluidStack syncing

krossgg committed:
Adds new IFluidHandler extensions

krossgg committed:
Switches IFluidTransfer for IFluidHandler

Rewrites these classes to use Forge's native IFLuidHandler rather than LDLib's IFluidTransfer

krossgg committed:
Changes fluid-related behavior in some classes

Switches from hardcoded logic to using Forge's FluidUtil to add cleanliness

krossgg committed:
Switched from LDLib FluidStack to Forge FluidStack

krossgg committed:
say the line bart

screret commented on src/main/java/com/gregtechceu/gtceu/api/capability/IThermalFluidHandlerItemStack.java:
```diff
@@ -26,12 +24,12 @@ public interface IThermalFluidHandlerItemStack {
         // fluids less than 120K are cryogenic
         if (temp < 120 && !isCryoProof()) return false;
-        if (FluidHelper.isLighterThanAir(stack) && !isGasProof()) return false;
+        if (fluidType.isLighterThanAir() && !isGasProof()) return false;
```
These should all use the fluidstack aware methods

krossgg replied:
`FluidType#getLighterThanAir` does not have a stack-aware version. 
Internally, this is Forge's implementation of the method: `return this.getDensity() <= 0;` 
In fact, the original LDLib FluidHelper method doesn't use a stack-aware version. 
This can be remedied if needed by doing the stack-aware version of the density and comparing it to 0.
I have changed it for any other property getters if applicable.

screret reviewed (changes requested):
I'm gonna do the rest in intellij github's UI isn't happy rn

krossgg committed:
Changed Drum and Tank right-click behavior

Now passes the interaction for drums and tanks if the held item isn't a fluidhandler
Also cleaned up the CreativeTankMachine to more properly make use of being a subclass
Implemented CreativeTankMachine right-click behavior

krossgg committed:
Changed fluid prop-getters to be stack-aware

screret commented on src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/properties/FluidPipeProperties.java:
```diff
@@ -103,8 +103,8 @@ public String toString() {
-    public long getPlatformThroughput() {
-        return getThroughput() * FluidHelper.getBucket() / 1000;
+    public int getPlatformThroughput() {
+        return getThroughput() * FluidType.BUCKET_VOLUME / 1000;
```
just change this to `return getThroughput()`

krossgg replied:
I have removed the `getPlatformThroughput` method entirely. It was never used anywhere (except in a file that's literally completely commented out but you can deal with that one).

screret commented on src/main/java/com/gregtechceu/gtceu/api/gui/widget/ScrollablePhantomFluidWidget.java:
```diff
@@ -19,9 +18,9 @@
 
     private static final int SCROLL_ACTION_ID = 0x0001_0001;
-    private static final long MILLIBUCKETS = FluidHelper.getBucket() / 1000;
+    private static final int MILLIBUCKETS = FluidType.BUCKET_VOLUME / 1000;
```
remove, this is useless.

screret commented on src/main/java/com/gregtechceu/gtceu/api/machine/steam/SimpleSteamMachine.java:
```diff
@@ -89,7 +89,7 @@ public void onLoad() {
         // Fine, we use it to provide eu cap for recipe, simulating an EU machine.
         capabilitiesProxy.put(IO.IN, EURecipeCapability.CAP,
-                List.of(new SteamEnergyRecipeHandler(steamTank, FluidHelper.getBucket() / 1000d)));
+                List.of(new SteamEnergyRecipeHandler(steamTank, FluidType.BUCKET_VOLUME / 1000d)));
```
just do `1d`

screret commented on src/main/java/com/gregtechceu/gtceu/api/machine/steam/SteamBoilerMachine.java:
```diff
@@ -198,13 +196,15 @@ protected void updateCurrentTemperature() {
                 if (hasDrainedWater) {
                     filledSteam = steamTank.fillInternal(
-                            GTMaterials.Steam.getFluid(fillAmount * FluidHelper.getBucket() / 1000), false);
+                            GTMaterials.Steam.getFluid(fillAmount * FluidType.BUCKET_VOLUME / 1000),
```
just pass `fillAmount`

screret commented on src/main/java/com/gregtechceu/gtceu/api/machine/steam/SteamBoilerMachine.java:
```diff
@@ -198,13 +196,15 @@ protected void updateCurrentTemperature() {
+                int fillAmount = (int) (getBaseSteamOutput() * ((float) currentTemperature / getMaxTemperature()) /
                         2);
-                boolean hasDrainedWater = !waterTank.drainInternal(FluidHelper.getBucket() / 1000, false).isEmpty();
+                boolean hasDrainedWater = !waterTank.drainInternal(FluidType.BUCKET_VOLUME / 1000, FluidAction.EXECUTE)
```
just pass `1` instead of `FluidType.BUCKET_VOLUME / 1000`

screret commented on src/main/java/com/gregtechceu/gtceu/api/machine/trait/FluidTankProxyTrait.java:
```diff
@@ -59,18 +55,17 @@ public int getTanks() {
     public void setFluidInTank(int tank, @NotNull FluidStack fluidStack) {
         if (proxy != null) {
-            proxy.setFluidInTank(tank, fluidStack);
+            // proxy.setFluidInTank(tank, fluidStack);
```
use IFluidHandlerModifiable so this can be overridden?

screret commented on src/main/java/com/gregtechceu/gtceu/api/machine/trait/NotifiableFluidTank.java:
```diff
@@ -177,14 +224,14 @@ public boolean isLocked() {
     public void setLocked(boolean locked) {
         if (this.isLocked() == locked) return;
-        FluidStack fluidStack = getStorages()[0].getFluid();
+        FluidStack fluidStack = storages[0].getFluid();
```
no, don't remove `getStorages()`, that method is overridden in some places, is it not?

krossgg replied:
It was overridden in one place for the CreativeTank for some stupid logic previously, however I have since removed that. It now uses the superclass member `storages` natively. It is no longer overridden anywhere.

screret commented on src/main/java/com/gregtechceu/gtceu/api/misc/lib/PhantomFluidWidget.java:
```diff
@@ -0,0 +1,269 @@
+package com.gregtechceu.gtceu.api.misc.lib;
```
api/gui/widget

screret commented on src/main/java/com/gregtechceu/gtceu/api/misc/lib/TagOrCycleFluidTransfer.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class TagOrCycleFluidTransfer implements IFluidHandlerModifiable {
```
name this `TagOrCycleFluidHandler`

screret commented on src/main/java/com/gregtechceu/gtceu/api/recipe/ingredient/FluidIngredient.java:
```diff
@@ -33,32 +32,32 @@ public class FluidIngredient implements Predicate<FluidStack> {
     public JsonElement toJson() {
         JsonObject jsonObject = new JsonObject();
-        jsonObject.addProperty("amount", this.amount * FluidHelper.getBucket() / 1000);
+        jsonObject.addProperty("amount", this.amount * FluidType.BUCKET_VOLUME / 1000);
```
just pass `this.amount`

screret commented on src/main/java/com/gregtechceu/gtceu/api/misc/lib/FluidTransferList.java:
move to api/transfer/fluid

screret commented on src/main/java/com/gregtechceu/gtceu/api/misc/lib/TagOrCycleFluidTransfer.java:
move to api/transfer/fluid

screret commented on src/main/java/com/gregtechceu/gtceu/api/misc/lib/TankWidget.java:
move to api/gui/widget

screret commented on src/main/java/com/gregtechceu/gtceu/common/cover/PumpCover.java:
```diff
@@ -62,15 +64,15 @@ public class PumpCover extends CoverBehavior implements IUICover, IControllable
             CoverBehavior.MANAGED_FIELD_HOLDER);
 
-    protected static final long MILLIBUCKET_SIZE = FluidHelper.getBucket() / 1000;
+    protected static final int MILLIBUCKET_SIZE = FluidType.BUCKET_VOLUME / 1000;
```
either remove this constant or just set it to 1.

krossgg replied:
Completely removed the constant, wherever it used.

screret commented on src/main/java/com/gregtechceu/gtceu/api/transfer/fluid/ModifiableFluidHandlerWrapper.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public void setFluidInTank(int i, FluidStack fluidStack) {
+        drain(handler.getFluidInTank(i), FluidAction.EXECUTE);
```
this should first simulate to check if insert/extract is allowed before executing.

krossgg replied:
fixed but also
https://github.com/GregTechCEu/GregTech-Modern/blob/a8bd4a86293bda85a622d9fe8f0c873ca016f394/src/main/java/com/gregtechceu/gtceu/api/transfer/fluid/ModifiableFluidHandlerWrapper.java#L17-L19
🧌

screret commented on src/main/java/com/gregtechceu/gtceu/common/item/ItemFluidContainer.java:
```diff
@@ -20,9 +20,9 @@ public ItemStack getRecipeRemained(ItemStack itemStack) {
         var storage = new ItemStackTransfer(itemStack);
         var transfer = FluidTransferHelper.getFluidTransfer(storage, 0);
         if (transfer != null) {
-            var drained = transfer.drain(FluidHelper.getBucket(), true);
-            if (drained.getAmount() != FluidHelper.getBucket()) return ItemStack.EMPTY;
-            transfer.drain(FluidHelper.getBucket(), false);
+            var drained = transfer.drain(FluidType.BUCKET_VOLUME, true);
+            if (drained.getAmount() != FluidType.BUCKET_VOLUME) return ItemStack.EMPTY;
+            transfer.drain(FluidType.BUCKET_VOLUME, false);
```
why does this still use the LDLib methods?

krossgg replied:
I forgor 💀 
Changed, but also I don't think this component even does anything LOL. I think it's a failed attempt at making fluid item crafting a reality but doesn't seem to work.

screret commented on src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/MultiblockTankMachine.java:
```diff
@@ -83,7 +87,7 @@ public Widget createUIWidget() {
 
     private String getFluidLabel() {
-        return String.valueOf(tank.getFluidInTank(0).getAmount() / (FluidHelper.getBucket() / 1000));
+        return String.valueOf(tank.getFluidInTank(0).getAmount() / (FluidType.BUCKET_VOLUME / 1000));
```
remove the division

screret commented on src/main/java/com/gregtechceu/gtceu/common/machine/storage/QuantumTankMachine.java:
```diff
@@ -344,7 +298,7 @@ public Widget createUIWidget() {
                 .addWidget(new LabelWidget(8, 8, "gtceu.gui.fluid_amount"))
                 .addWidget(new LabelWidget(8, 18,
-                        () -> String.valueOf(cache.getFluidInTank(0).getAmount() / (FluidHelper.getBucket() / 1000)))
+                        () -> String.valueOf(cache.getFluidInTank(0).getAmount() / (FluidType.BUCKET_VOLUME / 1000)))
```
remove useless division.

screret commented on src/main/java/com/gregtechceu/gtceu/common/machine/trait/FluidDrillLogic.java:
```diff
@@ -107,7 +106,7 @@ private long getFluidToProduce(FluidVeinWorldEntry entry) {
                 produced = produced * 3 / 2;
             }
-            return produced * FluidHelper.getBucket() / 1000;
+            return produced * FluidType.BUCKET_VOLUME / 1000;
```
remove useless division.

screret commented on src/main/java/com/gregtechceu/gtceu/integration/ae2/gui/widget/list/AEFluidDisplayWidget.java:
```diff
@@ -70,13 +72,13 @@ public void drawInForeground(@NotNull GuiGraphics graphics, int mouseX, int mous
                 if (!Platform.isForge()) {
                     tooltips.add(Component.literal(
-                            "§6mB:§r %d mB".formatted(fluidStack.getAmount() * 1000 / FluidHelper.getBucket())));
+                            "§6mB:§r %d mB".formatted(fluidStack.getAmount() * 1000 / FluidType.BUCKET_VOLUME)));
```
remove useless division.

screret commented on src/main/java/com/gregtechceu/gtceu/integration/ae2/gui/widget/list/AEFluidDisplayWidget.java:
```diff
@@ -70,13 +72,13 @@ public void drawInForeground(@NotNull GuiGraphics graphics, int mouseX, int mous
                 List<Component> tooltips = new ArrayList<>();
                 tooltips.add(fluidStack.getDisplayName());
-                tooltips.add(Component.literal(String.format("%,d ", fluid.amount())).append(FluidHelper.getUnit()));
+                tooltips.add(Component.literal(String.format("%,d ", fluid.amount())).append("mB"));
```
move the append into the String.format call itself, like: `String.format("%,d mB", ...)`

screret commented on src/main/java/com/gregtechceu/gtceu/syncdata/FluidStackLoad.java:
```diff
@@ -0,0 +1,34 @@
+
+import org.jetbrains.annotations.Nullable;
+
+public class FluidStackLoad extends ObjectTypedPayload<FluidStack> {
```
`FluidStackPayload`

screret commented on src/main/java/com/gregtechceu/gtceu/utils/GTUtil.java:
```diff
@@ -419,13 +424,65 @@ public static FluidStack getFluidFromContainer(Object ingredient) {
+     */
+    public static LazyOptional<IFluidHandler> getAdjacentFluidHandler(Level level, BlockPos pos, Direction facing) {
+        return FluidUtil.getFluidHandler(level, pos.relative(facing), facing.getOpposite());
+    }
```
move to GTTransferUtils?

screret commented on src/main/java/com/gregtechceu/gtceu/utils/GTUtil.java:
```diff
@@ -419,13 +424,65 @@ public static FluidStack getFluidFromContainer(Object ingredient) {
+        return FluidUtil.getFluidHandler(level, pos.relative(facing), facing.getOpposite());
+    }
+
+    // Same as above, but returns the presence
+    public static boolean isAdjacentFluidHandler(Level level, BlockPos pos, Direction facing) {
+        return getAdjacentFluidHandler(level, pos, facing).isPresent();
+    }
```
move to GTTransferUtils and rename to `hasAdjacentFluidTransfer`

krossgg committed:
Implemented Requested Changes

Removed any unnecessary BUCKET_VOLUME uses
Moved new classes to appropriate packages
Renamed Fluid classes to use Handler rather than Transfer

screret commented:
conflicts

krossgg committed:
Update LDLib overrides

Also clean up some lingering references

krossgg commented:
Last commit is the only new content. Rebased and resolved conflicts because things were getting a little out of hand.

krossgg committed:
spotless

Fuyukai commented:
This causes issues with loading certain machines: 

```
[16:54:12] [Server thread/ERROR] [minecraft/BlockEntity]: Failed to load data for block entity gtceu:mv_super_tank
java.lang.ClassCastException: class net.minecraft.nbt.LongTag cannot be cast to class net.minecraft.nbt.IntTag (net.minecraft.nbt.LongTag and net.minecraft.nbt.IntTag are in module minecraft@1.20.1 of loader 'TRANSFORMER' @84a9f65)
```

I don't know how much this actually affects (as none of my tanks emptied) but a few of my fluid pipes spontaneously disconnected.

krossgg committed:
Error handling for old LDLib tags

krossgg committed:
Remove warning

krossgg committed:
lesspot

krossgg committed:
Remove references to Transfer

krossgg commented:
Ready for merge after this last review. Will almost certainly break addons.

screret commented:
merge #1945 first

krossgg committed:
Merge from main with corrections

krossgg committed:
Migrate new things and touchups

krossgg committed:
Change widget registration priority so that custom UIs still work

krossgg committed:
More cleanup

krossgg committed:
Rework fluid recipe handling, again

krossgg committed:
Rework fluid jetpack to not bother with recipe searching

krossgg committed:
Spotless

screret commented on src/main/java/com/gregtechceu/gtceu/common/data/GTRecipeTypes.java:
```diff
@@ -463,7 +464,14 @@ public class GTRecipeTypes {
+                if (builder.input.containsKey(FluidRecipeCapability.CAP)) {
+                    Content input = builder.input.get(FluidRecipeCapability.CAP).get(0);
+                    FluidIngredient ingredient = FluidRecipeCapability.CAP.of(input.getContent());
+                    PowerlessJetpack.FUELS.put(ingredient, builder.duration);
```
This map should probably get cleared on reload before anything is added to it

krossgg committed:
Change where jetpack fuels get added for KJS compatibility
